experimental/dds/tree2/src/test/shared-tree/fuzz/fuzzEditGenerators.ts

```diff
@@ -308,114 +290,61 @@ export const makeEditGenerator = (
+		return contents === done
+			? done
+			: {
+					type: "edit",
+					contents,
+					index: state.treeIndex,
+			  };
 	};
-
-	return createAsyncGenerator<FieldEdit | NodeEdit, TreeEdit>(baseEditGenerator, buildOperation);
 };
 
 export const makeTransactionEditGenerator = (
 	opWeights: Partial<EditGeneratorOpWeights>,
-): AsyncGenerator<TransactionBoundary, FuzzTestState> => {
-	type EditState = FuzzTestState & TreeContext;
+): Generator<TransactionBoundary, EditState> => {
 	const passedOpWeights = {
 		...defaultEditGeneratorOpWeights,
 		...opWeights,
 	};
-	async function transactionStartGenerator(state: EditState): Promise<TransactionStartOp> {
-		return {
-			fuzzType: "transactionStart",
-		};
-	}
+	const start: TransactionStartOp = { fuzzType: "transactionStart" };
+	const commit: TransactionCommitOp = { fuzzType: "transactionCommit" };
+	const abort: TransactionAbortOp = { fuzzType: "transactionAbort" };
 
-	async function transactionCommitGenerator(state: EditState): Promise<TransactionCommitOp> {
-		return {
-			fuzzType: "transactionCommit",
-		};
-	}
-
-	async function transactionAbortGenerator(state: EditState): Promise<TransactionAbortOp> {
-		return {
-			fuzzType: "transactionAbort",
-		};
-	}
-
-	const baseTransactionEditGenerator = createWeightedAsyncGenerator<
-		FuzzTransactionType,
-		EditState
-	>([
-		[transactionStartGenerator, passedOpWeights.start],
+	const transactionBoundaryType = createWeightedGenerator<FuzzTransactionType, EditState>([
+		[start, passedOpWeights.start],
 		[
-			transactionCommitGenerator,
+			commit,
 			passedOpWeights.commit,
 			({ trees, treeIndex }) => transactionsInProgress(trees[treeIndex]),
 		],
 		[
-			transactionAbortGenerator,
+			abort,
 			passedOpWeights.abort,
 			({ trees, treeIndex }) => transactionsInProgress(trees[treeIndex]),
 		],
 	]);
 
-	const buildOperation = (contents: FuzzTransactionType, treeIndex: number) => {
-		const operation: TransactionBoundary = {
-			type: "transaction",
-			contents,
-			treeIndex,
-		};
-		return operation;
-	};
-
-	return createAsyncGenerator<FuzzTransactionType, TransactionBoundary>(
-		baseTransactionEditGenerator,
-		buildOperation,
-	);
-};
-
-function createAsyncGenerator<Op, OpOut>(
-	baseGenerator: (state: FuzzTestState & TreeContext) => Promise<Op | typeof done>,
-	buildOperation: (contents: Op, treeIndex: number) => OpOut,
-): AsyncGenerator<OpOut, FuzzTestState> {
-	return async (state: FuzzTestState): Promise<OpOut | typeof done> => {
-		const trees = state.trees;
-		// does not include last tree, as we want a passive client
-		const treeIndex = trees.length === 1 ? 0 : state.random.integer(0, trees.length - 2);
+	return (state) => {
+		const contents = transactionBoundaryType(state);
 
-		const contents = await baseGenerator({
-			...state,
-			treeIndex,
-		});
-		state.numberOfEdits += 1;
-		if (contents === done) {
-			return done;
-		}
-		return buildOperation(contents, treeIndex);
+		return contents === done
```

Comment:
one thing that's a bit annoying about this composition pattern is that things don't chain nicely; we have to break control flow here. Either making the typing more informative so that `transactionBoundaryType` is known to be an `InfiniteGenerator` (i.e. it can't return `done`) or introducing some nice monadic syntax could be a cool QoL improvement for this kind of thing.